### ISSUE TITLE
sonarjs/no-duplicate-string is triggered for flex-* strings.

### ISSUE THREAD

**Issue by PupoSDC** [SonarSource/eslint-plugin-sonarjs#227]
**I want to report a bug.**

**Reproducer**

react native styles rely on a series of constant strings that will often be duplicated in the same file, some of these trigger the `sonarjs/no-duplicate-string` rule. 

```javascript
// minimal reproducer if relevant
const styles = {
  firstStyle: { alignSelf: 'flex-start' },
  secondStyle: { alignSelf: 'flex-start' },
  thirdStyle: { alignSelf: 'flex-start' },
}
```

**Expected behavior**

no error

**Actual behavior**

```
  213:21  error  Define a constant instead of duplicating this literal 3 times
```

**eslint-plugin-sonarjs version:**     `"eslint-plugin-sonarjs": "0.7.0",`
**eslint version:**     `"eslint": "7.25.0",`
**Node.js version:** `    "node": "=14.16.1",`

**Rule key:** `sonarjs/no-duplicate-string`

**Comment by vilchik-elena** [comment]
Hello,

What is preventing you from creating a constant for `'flex-start'`?

**Comment by PupoSDC** [comment]
nothing prevents me, but it's not common practice in react-native projects. 
This value is also type safe when using typescript.

**Comment by vilchik-elena** [comment]
As you might see we are not really pro in various frameworks:) Could you describe the pattern which you believe should be ignored by the rule? Is it only TS applicable? How would you see exclusion list?

**Comment by PupoSDC** [comment]
It's hard to come up with a complete inclusive list for all libraries / frameworks...

I would suggest having an API similar to the `no-global-assign` that allows for exceptions ([see](https://eslint.org/docs/rules/no-global-assign)).

for my particular needs I could do 

```json
"sonarjs/no-duplicate-string": ["error", {"exceptions": ["flex-*"]}]
```

thanks for following this up!

**Comment by Anuovec** [comment]
In TS, you can create union types like:
`type Status = "Admin" | "User" | "Moderator"`

These are type-safe, therefore no mistyping, and cannot be changed. They behave similarly to constants.
The advantage is that you do not need to create another name for the same thing, e.g. `const ADMIN = 'admin'`

I believe these should be excluded from the rule.

**Comment by Anuovec** [comment]
As for what @PupoSDC wrote, that does not seem like a very good idea. Sure, it works in your particular case, but there are a lot of CSS properties, next time you might have `display: block` written numerous times. The list could get long very quickly. We should either find a pattern here or you can disable this rule completely.

**Comment by PupoSDC** [comment]
Well, give the power to the users ;)
In my projects I find this rule to be extremely valuable, but there are 2-3 exceptions that break it. I'd rather have the power to disable those than no power. Trying to predict all the exceptions for all the frameworks will lead to insanity very quickly for the sonar team.

**Comment by Anuovec** [comment]
In that case eslint gives plenty of power: https://eslint.org/docs/user-guide/configuring/rules#disabling-rules

However, https://github.com/facebook/react-native/blob/master/Libraries/StyleSheet/StyleSheetTypes.js#L374 React Native is using similar thing mention in my comment above. UNfortunately, they are using flow not TS.

**Comment by Slapbox** [comment]
I think a whitelist is very important to make the most of this rule. I saw we had a bunch of duplicate strings, but most of them are HTML attributes. Making these into constants makes the code less readable as the person reading has to make sure they know the value of the constant, whereas they might know the HTML attribute off the top of their head.

**Comment by Pantalaimon** [comment]
I’m upgrading an old extjs project with this linting rule and have similar issues.

We have some ugly copy pasta duplicates, but also lots I consider reasonable :
- extjs components "xtype" like "x-checkbox". They may even be dangerous to factor into a constant (now I want to change my checkbox into a combo… so I need to change the xtype, which is defined here… there we go!)
- i18n message keys, like `getMsg('whatever.key')` and `getMsg('common.prefix.'+suffix)`. Factoring into constant introduce an extra indirection between the translated text (`whatever.key=Hello`) and its usage (`label : …`) which is annoying when we want to find the code associated to the button with label xxx.

Enforcing the rule on tens of thousands of lines of legacy code is tedious for a marginal maintainability gain. I would love to provide a "ignorePattern" config to skip `[a-z-]+` and `[a-z.]+`.

**Comment by PupoSDC** [comment]
Closing as I'm no longer working with this extension.

**Comment by Slapbox** [comment]
@PupoSDC thanks for your great work; I'll be sad to see such a wonderful plugin stop being developed but I understand that open source can be quite time consuming.

**Comment by PupoSDC** [comment]
ah, thanks a lot @Slapbox, but what I meant was that I am not working WITH this extension as a "customer". I never contributed directly to this project myself. 

Have a great day!

**Comment by safakyavsan** [comment]
> Closing as I'm no longer working with this extension.

What is your advice rather than this?

**Comment by PupoSDC** [comment]
Hey, no advice. 

This was a feature request from me (im not a maintainer or a developer), since this didn't get any traction, I no longer need it, and it was just polluting my feed i closed it.
